### PR TITLE
[BEAM-2271] Exclude files not intended to be in the releases source distribution

### PR DESCRIPTION
**Pull request by jbonofre**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [X] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [X] Each commit in the pull request should have a meaningful subject line and body.
 - [X] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [X] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [X] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [X] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by jbonofre** [comment]
It's possible to mimic what the release plugin is doing using:

```
mvn clean install -Prelease,apache-release
```

Then you can check the resulting source zip file in the target folder.

File: src/assembly.xml [hunk 1]
```diff
@@ -0,0 +1,111 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (iemejia) [inline src/assembly.xml]: We should avoid introducing a new directory 'src' in the current structure, please move it probably into the parent directory.

Comment (jbonofre) [reply src/assembly.xml]: I didn't put there but it's not related to the Java SDK. It's more generic and inherited from the Apache POM.

Comment (iemejia) [reply src/assembly.xml]: well then put it in the parent directory but wait before re pushing I have some some secs I have some extra things I found.

File: src/assembly.xml [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+
+                <!-- files generated by the Gradle build process. -->
+                <exclude>**/.gradle/**/*</exclude>
+                <exclude>**/.gogradle/**/*</exclude>
+                <exclude>**/build/**/*</exclude>
+                <exclude>**/vendor/**/*</exclude>
+                <exclude>**/.gradletasknamecache</exclude>
+
+                <!-- files generated by the Maven build process. -->
+                <exclude>**/bin/**/*</exclude>
```
Comment (iemejia) [inline src/assembly.xml]: When I run the command the bin filter is not being excluded, don't know why can you validate that it is currect ?

Comment (jbonofre) [reply src/assembly.xml]: Not sure I follow you there. What do you mean ?

File: src/assembly.xml (hunk 1 above)
Comment (jbonofre) [reply src/assembly.xml]: What `parent` directory ? I saw that a `release` folder has been created, I can eventually put there.

Comment (jbonofre) [reply src/assembly.xml]: OK, I move the assembly in the root folder.

File: src/assembly.xml (hunk 2 above)
Comment (jbonofre) [reply src/assembly.xml]: OK you mean exclude the folder itself (not the content).

File: assembly.xml [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+
+                <!-- file generated for archetypes -->
+                <exclude>sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/</exclude>
+                <exclude>sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/src/</exclude>
+
+                <!-- files generated by the Python build process. -->
+                <exclude>**/*.py[cod]</exclude>
+                <exclude>**/*.egg-info/</exclude>
+                <exclude>**/.eggs/</exclude>
+                <exclude>**/nose-*.egg/</exclude>
```
Comment (iemejia) [inline assembly.xml]: directories finished in .egg should be excluded too.

File: assembly.xml [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                <exclude>**/dependency-reduced-pom.xml</exclude>
+
+                <!-- file generated for archetypes -->
+                <exclude>sdks/java/maven-archetypes/examples/src/main/resources/archetype-resources/src/</exclude>
+                <exclude>sdks/java/maven-archetypes/examples-java8/src/main/resources/archetype-resources/src/</exclude>
+
+                <!-- files generated by the Python build process. -->
+                <exclude>**/*.py[cod]</exclude>
+                <exclude>**/*.egg-info/</exclude>
+                <exclude>**/.eggs/</exclude>
```
Comment (iemejia) [inline assembly.xml]: This one seems not to be working, it inlcudes the python.eggs subdir with its children *.egg directories

Comment (jbonofre) [reply assembly.xml]: OK, let me try.

Comment (jbonofre) [reply assembly.xml]: Which file did you check ? `apache-beam-2.3.0-SNAPSHOT-source-release.zip` or `apache-beam-2.3.0-SNAPSHOT-src.zip` in the target folder ?

Comment (jbonofre) [reply assembly.xml]: The `src` is fine, the `source` (coming from the Apache POM) is not. Let me exclude this one.

**Comment by jbonofre** [comment]
FYI, it's the approach I use in Apache Karaf and other project to override the assembly. Basically, we don't use the `apache-release` profile: instead we use our own `release` profile doing the same but overriding the assembly execution.

File: assembly.xml [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                <exclude>%regex[(?!((?!${project.build.directory}/)[^/]+/)*src/)(.*/)?\.settings(/.*)?]</exclude>
+                <exclude>%regex[(?!((?!${project.build.directory}/)[^/]+/)*src/)(.*/)?\.externalToolBuilders(/.*)?]</exclude>
+                <exclude>%regex[(?!((?!${project.build.directory}/)[^/]+/)*src/)(.*/)?\.deployables(/.*)?]</exclude>
+                <exclude>%regex[(?!((?!${project.build.directory}/)[^/]+/)*src/)(.*/)?\.wtpmodules(/.*)?]</exclude>
+
+                <!-- misc -->
+                <exclude>%regex[(?!((?!${project.build.directory}/)[^/]+/)*src/)(.*/)?cobertura\.ser]</exclude>
+
+                <!-- files generated by the Gradle build process. -->
+                <exclude>**/.gradle</exclude>
```
Comment (iemejia) [inline assembly.xml]: Please add missing '/' at the end in this line and the next 4 (until vendor).

File: assembly.xml [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+                <exclude>**/.gogradle</exclude>
+                <exclude>**/build</exclude>
+                <exclude>**/vendor</exclude>
+                <exclude>**/.gradletasknamecache</exclude>
+
+                <!-- files generated by the Maven build process. -->
+                <exclude>**/bin</exclude>
```
Comment (iemejia) [inline assembly.xml]: Please add missing '/' at the end in this line.

**Comment by iemejia** [comment]
Please self merge after the last fixes.
One last thing I noticed are two 'new' directories that contain binary files:
sdks/java/extensions/sql/src/main/codegen
sdks/python/apache_beam/testing/data/vcf
Not sure if they should be exluded or no from the release. Should they?

**Comment by jbonofre** [comment]
In SQL `codegen` is generated by hand, not by the build, so it has to be part of the release. Same for `data/vcf`.

**Comment by jbonofre** [comment]
I will merge after a Jenkins run. Thanks !

**Pull request merged**